### PR TITLE
Adding PUPPI combined isolation Working Points in miniAOD muon selector

### PR DESCRIPTION
**Pull request by bmahakud**
#### PR description:
- New PR made after closing #26076
- Adding puppi combined isolation working points in muon selectors
- 3 working points added: Loose, Medium and Tight 
- No year dependence 
- All studies and discussion could be found in the ticket: https://its.cern.ch/jira/browse/CMSMUONS-128

### CODE REVIEW

**Comment by bmahakud** [comment]
@slava77 
All your comments have been addressed from the PR # 26076
except : 
"I think that it's more practical to use PFCandidate::ParticleType and convert to it using PFCandidate::translatePdgIdToType"  because the developers had defined  their own categories to decide which particles will be in the isolation calculation  and I do not want to use an inbuilt function. If it is not a big problem  may be we can proceed with this.

**Comment by slava77** [comment]
@cmsbuild please test

File: PhysicsTools/PatAlgos/plugins/PATMuonProducer.cc [hunk 1]
```diff
@@ -815,6 +831,82 @@ double PATMuonProducer::getRelMiniIsoPUCorrected(const pat::Muon& muon, float rh
...
+  if( id == 321    ) return true ;
+  if( id == 999211 ) return true ;
+  if( id == 2212   ) return true ;
+  return false;
+}
+
+bool PATMuonProducer::isPhoton( long pdgid ){
+  const long id = abs( pdgid );
+  if( id == 22 ) return true ;
+  return false;
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATMuonProducer.cc]: There is no ```-22```
replace the body with 
```return pdgid == 22;```

File: PhysicsTools/PatAlgos/plugins/PATMuonProducer.cc [hunk 2]
```diff
@@ -815,6 +831,82 @@ double PATMuonProducer::getRelMiniIsoPUCorrected(const pat::Muon& muon, float rh
...
+
+}
+
+bool PATMuonProducer::isChargedHadron( long pdgid ){
+  const long id = abs( pdgid );
+  if( id == 211    ) return true ;
+  if( id == 321    ) return true ;
+  if( id == 999211 ) return true ;
+  if( id == 2212   ) return true ;
+  return false;
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATMuonProducer.cc]: we are dealing with pf/packed candidates and there is only one ID per CH.
replace the body with ```return std::abs(pdgid) == 211;```

File: PhysicsTools/PatAlgos/plugins/PATMuonProducer.cc [hunk 3]
```diff
@@ -815,6 +831,82 @@ double PATMuonProducer::getRelMiniIsoPUCorrected(const pat::Muon& muon, float rh
...
+  return reliso_Puppi_combined;
+}
+
+bool PATMuonProducer::isNeutralHadron( long pdgid ){
+  const long id = abs( pdgid );
+  if( id == 111 ) return true ;
+  if( id == 130 ) return true ;
+  if( id == 310 ) return true ;
+  if( id == 2112 ) return true ;
+  return false;
```
Comment (slava77) [inline PhysicsTools/PatAlgos/plugins/PATMuonProducer.cc]: we are dealing with pf/packed candidates and there is only one ID per NH.
replace the body with ```return std::abs(pdgid) == 130;```

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by fabiocos** [comment]
@bmahakud the unit test error is unrelated to this PR, it is present in the IB as well

**Comment by slava77** [comment]
In JetHT dataset (wf 136.8311)  it looks like the new isolation flags are filled and a large fraction passes  all of them [the part of the red histogram not displayed is above 29.0 / PuppiIsoTight]
all_OldVSNew_RunJetHT2017FreMINIAODwf136p8311c_log2max0_5,patMuons_slimmedMuons__PAT_obj_selectors

Looking at the values manually, it seems like the PUPPI iso loose/med/tight bits are fairly well correlated with the PFIso bits. This is my superficial validation of the new bits.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/26233	61a19b6
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show differences only in the slimmedMuons selectors flags, corresponding to the newly filled bits.

**Comment by slava77** [comment]
@bmahakud 
Do we have any monitoring of the puppi isolation for muons in the standard DQM/validation?
If not, I suggest that it is added so that potential problems can be detected with the standard tools.

**Comment by slava77** [comment]
@fabiocos 
please check this
in case this is not on your list due to the analysis signature

**Comment by fabiocos** [comment]
please test

**Comment by santocch** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**